### ISSUE TITLE
SFZ Player - only noise output with various SFZ files

### ISSUE THREAD

**Issue by brylie** [squinkylabs/SquinkyVCV#231]
I have configured the SFZ Player in a simple patch, as follows:

![image](https://user-images.githubusercontent.com/17307/131109764-2711d3bf-d9fb-428a-8c9b-4e70d215809d.png)

When I load any SFZ file into the patch, it only produces a steady stream of noise. I have tried with [Versilian Studios Upright 1](https://vis.versilstudios.com/upright-1.html), [FreePats Acoustic Grand](https://freepats.zenvoid.org/Piano/acoustic-grand-piano.html) (both FLAC and WAV).

I am running VCV Rack on Ubuntu Linux.

**Comment by brylie** [comment]
I am downloading SFZ Player from the VCV Library.

FWIW, running Ubuntu within [VirtualBox](https://www.virtualbox.org/) on MacOS or Windows will have sound output.

**Comment by brylie** [comment]
Also, the noise isn't actually audible. It just shows as maxed-out VU meter and _only_ when the v/Oct is plugged in:


With V/Oct:
![image](https://user-images.githubusercontent.com/17307/131304360-8b1a632e-6966-4697-bfc2-6873eca821b3.png)


Without V/Oct:

![image](https://user-images.githubusercontent.com/17307/131304282-fcdda8e8-3f22-458c-bb2b-34065c2d2ff2.png)

**Comment by burkingman** [comment]
I'm having a similar issue. Using SFZ Player straight from the VCV Rack library, in Ubuntu Studio 21.04. VCV Rack overall works fine.

Tried building a very basic patch: MIDI-CV sending gate, velocity and V/oct from the computer keyboard to SFZ Player, with the SFZ player output going to Scope and to NYSTHI Multivoltimetro for diagnostic purposes (I also put in an Audio-8 module to complete the patch, but no sound is sent to it). Loaded the K18 upright piano into SFZ Player, then tried playing it. The first note plays fine, but the moment I release the key, SFZ Player starts outputting NaN and stays stuck that way, even if I unplug its inputs.

Here are the few possibly-relevant lines I could find in the log file:
- - -
[11,245 info ./dsp/samp/SamplerErrorContext.h:15] unimplemented opcodes:
[11,245 info ./dsp/samp/SamplerErrorContext.h:18] off_by
[11,245 info ./dsp/samp/SamplerErrorContext.h:18] on_hicc64
[11,245 info ./dsp/samp/SamplerErrorContext.h:18] on_locc64
- - -

Anybody else? According to [this comment](https://community.vcvrack.com/t/has-anyone-success-with-sfz-player-on-linux/13954/2), the problem may not be specific to SFZ Player, but I've only encountered it with this module so far.

**Comment by brylie** [comment]
Here are two screenshots before and after restarting Rack.

Prior to restarting Rack, there are indeed some `-nAn` values.

![nan values](https://community.vcvrack.com/uploads/default/optimized/2X/5/5d9226ef40b9b0a207bc60ca524151b53a64fb0f_2_416x500.png)

After restarting Rack, the `nAn` values are replaced by floats, but the SFZ player will only play one note.

![after restart](https://community.vcvrack.com/uploads/default/optimized/2X/4/4de6d96eed2adc13193283691a95590df429d8ab_2_410x500.png)

**Comment by burkingman** [comment]
Interesting. The first screenshot from @brylie in the post above shows MIDI-CV outputting NaN, but as far as I can tell, it's not the cause. I tried sending constant values to the velocity and V/oct inputs on SFZ Player (4V and 1V respectively, generated by NYSTHI's Const Add Mult module), and used the basic Pulses module to send a gate to the player. A note plays, but as soon as the gate goes down, the player gets stuck outputting NaN. All values sent to the SFZ Player are valid and remain valid once the bug acts up, but the player outputs NaN. This may well be an underlying VCV Rack problem, but if so, it doesn't stem from the MIDI CV module specifically.

**Comment by squinkylabs** [comment]
I am unable to get a NaN out of MIDI-CV when I run on Ubuntu 21. I did a straight download, and installed all my plugins that I had previously selected.

Could someone who has this issue (Nan out of MIDI-CV) provide more details? It is not a common problem at all.

**Comment by burkingman** [comment]
What kind of details would help? I included error messages in my first comment, but I don't know if there's anything useful there. I tried running Rack from the command line to see its output, but it doesn't output any message when the problem occurs.

I ran more tests using a very simple "player" patch that sends constants to the SFZ Player's velocity and V/oct inputs, and sends a gate to the player using the VCV Pulses module. If I have VCV Rack open that patch upon launch, SFZ Player works great, but it only works at launch. If I then re-open the patch and try to play a note, I get NaN. If I launch VCV Rack with any patch, even a practically empty one, then try running my SFZ Player patch, I get NaN. (I often use Jack, but ran the latest test using ALSA instead. As far as I can tell, the use of Jack or ALSA doesn't seem to make a difference.)

I did not use the MIDI-CV module in any of the latest tests; it doesn't seem to be the cause of the problem. I'm still trying to narrow it down further... Let me know if there's anything else you'd like me to try.

**Comment by squinkylabs** [comment]
well, I know midi-cv isn't the cause of the problem when you don't use it (I do read these - even if I was away). The reason I mention it is that is seems on this system MIDI-CV produces nans, and so does SFZ player. But for most people neither of these modules do. The only common thing is your system.

Since I don't have a system that shows any problems, I don't know what more I can do.

Is this an accurate summary of the situation? I have to confess there is a lot going on in this thread, so it's difficult for me to follow the various issues.

**Comment by squinkylabs** [comment]
@brylie : do you have anything definitive? Are you also seeing both MIDI-CV and SFZ player outputting NaN? 

Has anyone tried building the modules in question themselves and running them?

If this bug can't be reproduced on mac and windows ever, and on linux only on a small percentage of systems, it will be very, very difficult to find.

Has anyone seen a situation where MIDI-CV does work correctly and SFZ Player doesn't?

btw: SFZ player (like many polyphonic modules) gets its polyphony from the number of channels on the input. When that number is zero (like no patchcord), then SFZ player does absolutely nothing. So it isn't surprising that the issue goes away.

**Comment by squinkylabs** [comment]
If anyone could upload a patch that shows the problem, I could try that one my different linux VMs. I think I have 18, 20, and 21?

**Comment by brylie** [comment]
@squinkylabs I'll follow up ASAP. Thank you so much for your help so far :smile: For what it's worth, I was able to successfully use SFZ Player on an [experimental patch](https://patchstorage.com/escher-adams-cellular-automata-piano/).

**Comment by squinkylabs** [comment]
ok, thanks for being patient. If there is a bug in SFZ player around this I want to fix it ;-)

**Comment by squinkylabs** [comment]
@burkingman : OK, just noticed the part about re-opening a patch. It would be super, super, super useful if you could send me a patch that has this problem.

**Comment by burkingman** [comment]
Thanks for the precisions about MIDI-CV. I'll try testing it separately, I'm curious to see if I can get NaN out of it. So far, when I test it with SFZ Player, it behaves normally even when SFZ Player doesn't (unlike what seems to happen on @brylie's system).

[Here's my basic test patch for the SFZ Player](https://ericgauthier.net/fichiers/sfz_player_test_from_burkingman.vcv) (temporary link, but I'll keep it available while this issue is open). Unless I have VCV Rack open the patch upon launch, SFZ Player outputs NaN after the first note (as soon as I release the button on Pulses).

I've used and enjoyed many of your plugins and haven't had any problems until now. I tried downloading them again (by deleting them, then unsubscribing and resubscribing in the VCV Library). Doesn't seem to change anything. Haven't tried building them myself -- I usually don't build anything from source, though I may get around to trying it.

**Comment by squinkylabs** [comment]
@burkingman thanks - I'll try this out.

what does this mean: " Unless I have VCV Rack open the patch upon launch, SFZ Player outputs NaN after the first note (as soon as I release the button on Pulses)." does that mean the steps are 1) launch VCV 2) load patch, 3) press button -> observe Nan? Sorry to be pedantic, but these reports are very difficult to keep straight.

btw: I assumed most linux users build themselves. It's so hard to get a system that's compatible with pre-build binaries. But maybe we will get lucky and i can repro this.

Unfortuantely for you, Linux is a miniscule percentage of the rack user base, so I rarely have linux testers, and when I do they always build themselves.

Just tried you test file under windows. Of course it works fine there. I'll try my ubuntu 20 next, but that had a VCV I built myself. I will be difficult for me to test in my ubuntu 21 VM, as that one has no shared files. But I can try...

**Comment by squinkylabs** [comment]
@burkingman ok, I configured file shaing for my linux 21 VM. I can load your patch fine, and when I press the first tap button it does what I expect, and I don't see any NaN.

But at least my linux 21 with pre-built rack is working now, so that's nice.

**Comment by burkingman** [comment]
> what does this mean: " Unless I have VCV Rack open the patch upon launch, SFZ Player outputs NaN after the first note (as soon as I release the button on Pulses)." does that mean the steps are 1) launch VCV 2) load patch, 3) press button -> observe Nan? Sorry to be pedantic, but these reports are very difficult to keep straight.

No problem, troubleshooting requires precision. The steps are as you describe, yes. The only way I can get this patch to work normally (no NaN) is by launching VCV from the command line with the patch filename as an argument: then VCV opens the patch upon launch and it works as it should.

Thanks for your efforts in trying to reproduce the problem. Maybe it has to do with my system configuration... I'll let you know if I think of anything else or have any new data.

**Comment by squinkylabs** [comment]
ok, thanks. I'll keep my eyes open, too.

**Comment by squinkylabs** [comment]
@burkingman : I wonder if it could be this? (stack corruption by other code) https://community.intel.com/t5/Intel-C-Compiler/Unexpected-NaN-as-return-value/td-p/992170

**Comment by brylie** [comment]
> I assumed most linux users build themselves.

Linux is increasingly accessible to computer users who may not be as familiar with software build tools. There are many mainstream Linux distributions, such as Debian, Ubuntu, and OpenSUSE, that provide extensive repositories of pre-built software. While the average Linux user may be likely to have build tools installed, it isn't necessarily due to personal preference. :smiley:

**Comment by brylie** [comment]
By the way, @squinkylabs, do you have a "buy me a coffee" or GitHub sponsors account set up?

**Comment by brylie** [comment]
Here are my steps to reproduce:

1. create a patch with V/Oct  and Gate sources
2. connect the sources to SFZ Player and output to Volt Meter
3. trigger the gate (nothing abnormal happens)
4. **load an instrument in SFZ Player**
5. trigger the gate (note the `-nAn`)

https://user-images.githubusercontent.com/17307/132301030-47a65ed2-4986-4cc9-9f11-770b5e6dc5b5.mp4

This happens reliably regardless of whether I reloaded VCV Rack with the patch. I.e. the bug manifests immediately after loading the SFZ instrument into SFZ player.

**Comment by brylie** [comment]
Is there some way I can provide more detailed logs or stack trace?

**Comment by squinkylabs** [comment]
@brylie : I don't have any way to give me things, sorry. I do this for fun, and have a "job" to make money, so it works out fine.

what do you use to print out your linux info like that? (I know very little about Linux). 

How would I install something "just like" what you have? Is there a different distro with a realtime kernel, or is that something you install on top of regular ubuntu?

Please, please, please provide a patch along with bug reports.

I could probably make a special version that logged some info around this issue, but the normal version won't have any useful logging. That said, have you looked at the VCV log file right after this happens? It's possible (but not super likely) that there will be info there. Making a new version might be a good way to go.

btw, have you used the VCV ADSR module in this setup? I took code from there to make the envelope generators in SFZ Player.

**Comment by squinkylabs** [comment]
re other message - sure, no one likes to build stuff. I just didn't think Linux was to the point of user friendliness that music software always works fine out of the box. I, too, was pleasantly surprised how good `apt get` is. Although last time I used Linux at work I still needed to build a ton of (admittedly esoteric) stuff.

**Comment by squinkylabs** [comment]
Actually, as I ponder making a build with debugging stuff in it, I should post a "developer build" first. I suspect that will make the problem go away. Which won't help us find it, but will be interesting. I'll do that this evening, if not sooner.

**Comment by squinkylabs** [comment]
@brylie : Here is a linux "dev build". Could you try it and tell me if you still have the problem?
[squinkylabs-plug1-1.0.22-lin.zip](https://github.com/squinkylabs/SquinkyVCV/files/7124768/squinkylabs-plug1-1.0.22-lin.zip)

**Comment by brylie** [comment]
Thanks, @squinkylabs I will try ASAP. 

For what it's worth, I was not able to reproduce the issue on a clean installation of Kubuntu (basically Ubuntu with KDE). I used a bootable USB with Kubuntu to install Rack and these plugins. Then I created the basic "reproduction" patch, which worked correctly.

I suppose there may be some issue related to installed software, such as the kernel version, NVIDIA video drivers, or some other development tool/library that I may have installed along the way. By the way, I only installed the real-time kernel a few days ago, after experiencing this issue with a different kernel version.

**Comment by burkingman** [comment]
> @burkingman : I wonder if it could be this? (stack corruption by other code) https://community.intel.com/t5/Intel-C-Compiler/Unexpected-NaN-as-return-value/td-p/992170
Hard to tell. I'd be inclined to think it's not, considering how old those posts are, but you never know...

I just applied a minor kernel upgrade and tried your dev build (thanks!) and I get the same bug.

What @brylie's system and mine have in common is Kubuntu running on an Intel i5, though we're on slightly different versions of the OS and kernel (I'm running Ubuntu Studio, which is basically Kubuntu optimized for multimedia production). Interesting that a fresh Ubuntu and even a fresh Kubuntu don't have the same problem. So, yeah, system config could be the cause, somehow. Earlier today I noticed a bug in a module from another developer; wondering if it could be related. I'll report back if I find anything useful.

**Comment by squinkylabs** [comment]
@burkingman I'm glad that you still have the error with my build. That means I can make a build with more checks in it and maybe we can find something out.. it seems pretty clear there is something not quite right with you systems, as far as VCV compatibility. It seems like you have issues with multiple modules. 

I'm happy to experiment around to find what it is, but I really doubt that it's going to be super enlightening.

**Comment by squinkylabs** [comment]
@burkingman and @brylie : here is a new dev build. This one is designed to crash (actually assert) if it sees a NaN. The hope is that this will point me to where these NaNs are coming from. If you run this build and do any of the normal Nan things it will crash, but it will leave behind information in the log file. Most importantly where it found the NaN. So if you get this to crash, get the last part of the log file!

This build is not even a full plugin zip (don't ask, I have my reasons). It's just the plugin.so in a zip file. Unzip it and copy of on top of the one that is normally in the SquinkyLabs folder.

[plugin.zip](https://github.com/squinkylabs/SquinkyVCV/files/7133672/plugin.zip)

**Comment by burkingman** [comment]
Thanks for this! Is it meant to be so much bigger than the regular plugin.so file? (over 60 MB vs 3.5 MB)

**Comment by squinkylabs** [comment]
@burkingman haha - yes, wondering if you would notice. the normal `make dist` runs strip on the so and makes is much smaller. Dev build `make` doesn't do that. Either would be ok for this, probably.

**Comment by burkingman** [comment]
OK, just tried it: my test patch crashes as soon as I hit the button to start playing a note (usually, the problem occurred when releasing the button). I'm attaching the log (I took out a buch of lines about loading SVG files):
[sfz_player_crash_log.txt](https://github.com/squinkylabs/SquinkyVCV/files/7140908/sfz_player_crash_log.txt)

**Comment by squinkylabs** [comment]
@burkingman - ok that helps. I'll make one more more details in that area. maybe next time could you run rack from a terminal session? I expect to see something like this in the console:

```assertion at line 170: std::isnan()```

I'll try to add more debugging in a way that will log more good stuff so you don't need to worry....

**Comment by squinkylabs** [comment]
[plugin.zip](https://github.com/squinkylabs/SquinkyVCV/files/7141425/plugin.zip)
 new debug dev build. should give (a little bit?) more information, but I could really use the console output. Especially if it has line numbers like the message, above.

**Comment by brylie** [comment]
Here is a screen recording of the bug with console output from the plugin.zip you provided above.


https://user-images.githubusercontent.com/17307/132807886-03fc8ea0-18ea-4b37-a94e-ab498cb50f29.mp4

**Comment by squinkylabs** [comment]
@brylie thanks! So - it's the ADSR! when I'm squinking tonight I'll see if I can fix it, or it I still need another round of test the focus on the suspects.

btw, that command line output in your video is exactly what I need (although the log file would probably have done in this case. If you need to test again, that one line for where the assert happened (in this case setLambda line xxx) it exactly what I need.

**Comment by squinkylabs** [comment]
OK, I see what's happening (I think), but not exactly why. This build should tell more. I added a LOT of logging, so I need all the data after the section on unimplemented opcodes. It will be hundreds of lines like this:
```
[0.465 info dsp/samp/CompiledRegion.cpp:104] reading region release = 0.500000
[0.465 info dsp/samp/CompiledRegion.cpp:104] reading region release = 0.030000
[0.465 info dsp/samp/CompiledRegion.cpp:104] reading region release = 0.500000
[0.465 info dsp/samp/CompiledRegion.cpp:104] reading region release = 0.500000
```

Followed by some other stuff. probably a zipped up log file would do, along with the assert info (line number and file).

[plugin.zip](https://github.com/squinkylabs/SquinkyVCV/files/7144773/plugin.zip)

**Comment by burkingman** [comment]
Sorry, should have thought to run from console. Tried your new build and got this error in the console:

> Rack: ./composites/ADSRSampler.h:119: void ADSRSampler::setRSec(float): Assertion `t > 0' failed.
Aborted (core dumped)

Here's the log, from the loading of the patch:
[sfz_player_crash_log2.txt](https://github.com/squinkylabs/SquinkyVCV/files/7144868/sfz_player_crash_log2.txt)

**Comment by squinkylabs** [comment]
@burkingman oh! very interesting. The adsr is getting set to a release time of zero seconds, and setLambda divides by that, giving infinity. I see several of the regions are getting release time zero.

It looks like on your system the SFZ region is not correctly inheriting the release value from the enclosing group. Which is pretty weird, but clearly happening.

I never suspected a Linux error in SFZ parsing, but that seems to be what is happening. Will need to do some more serious thinking this evening.

I can easily patch the adsr to avoid the NaN and other craziness, but it the SFZ parser is not seeing groups it's never going to work super well. More to come.

**Comment by squinkylabs** [comment]
I'm sure this is pretty dull from the outside, but since it's looking like some kind of error in the SFZ parser of my own creation it's interesting to me. Not what I was expecting, and presumable easy to fix once we find what the cause is.

**Comment by burkingman** [comment]
> I'm sure this is pretty dull from the outside

That's all right, I don't come to GitHub for the entertainment. :) By the way, I tested with a simpler SFZ file ([this one](https://musical-artifacts.com/artifacts/1471)) and that worked well. I'd been using the K18 piano for my tests since it was listed in the manual for SFZ Player, but I suppose there's more to parse with that one, and that triggers the bug.

**Comment by squinkylabs** [comment]
@burkingman yeah, but that's one I've used more than any other. I have quite a complete suite of parser test that all run fine. Clearly something is wrong.. I'm working on porting all the parser tests to linux. will get to the bottom of this!

**Comment by squinkylabs** [comment]
This version has a lot of logging in it. About 6000 lines of text will come out when you load piano 18. It will probably crash/assert much earlier, too. In this case I don't care that much about the line it asserts on, it's mostly the thousands that came first. so save off a log file, zip it and post it.

[plugin.zip](https://github.com/squinkylabs/SquinkyVCV/files/7148552/plugin.zip)

**Comment by burkingman** [comment]
All right, here's the new test log, starting from the loading of the patch: 
[sfz_player_crash_log3.zip](https://github.com/squinkylabs/SquinkyVCV/files/7149714/sfz_player_crash_log3.zip)

And the relevant console output:

> Rack: dsp/samp/CompiledRegion.cpp:105: void CompiledRegion::addRegionInfo(SamplerSchema::KeysAndValuesPtr): Assertion `ampeg_release > 0' failed.
Aborted (core dumped)

**Comment by squinkylabs** [comment]
Thanks, @burkingman. Looking forward to going through this. As you can see, I moved the error checking farther up-stream, to the code that "compiles" a playable patch from the output of the parser. I will remove that error check in the future, because I think ampeg_release zero is actually valid. It's just not expected with the piano patch.

**Comment by squinkylabs** [comment]
hey, @burkingman and @brylie : do you have your Ubuntu set to some locale where the number I call "point zero three" is written as `0,03`, rather than the US convention of `0.03` (period vs coma used as a decimal point)? I'm getting a little suspicious....

Also, if so, could you a) tell me if it fixes it to set a US locale, and b) how can I set this in my ubuntu?

**Comment by burkingman** [comment]
Ah ha! My system does indeed display "0.03" as "0,03". If I set it to US number formats, the SFZ Player works well and I can't trigger the bug.

Not sure how you'd change it in Ubuntu: it depends which flavour you're running. I had to open System Settings, then look under Regional Settings, in the Formats subsection.

**Comment by squinkylabs** [comment]
OK, I'll try that. I figure now that I know what it is I might as well make it work in all locales. Thanks for sticking with it.

btw - I changed my test suite to switch to Italian before running the tests. Now I get massive failures ;-)

**Comment by brylie** [comment]
> do you have your Ubuntu set to some locale where the number I call "point zero three" is written as 0,03, rather than the US convention of 0.03 (period vs coma used as a decimal point)?

Yes. My locale is `fi_FI`, which uses a decimal comma.

**Comment by squinkylabs** [comment]
I am unable to duplicate this issue in Linux, even if I set my number format to `,`. Oh well, it's always fun. I'll make a Linux build tonight with the fixes. I suspect it will fix these issues.

I wonder if the issues with MIDI-CV outputting nan is related to non-us keyboard / locale?

**Comment by squinkylabs** [comment]
Here's version that is hopefully fixed, but as I said above I can't really test myself. This time it's a "normal" build and contains the full plugins. Tell me if it works, and post a log file, please.

[squinkylabs-plug1-1.0.22-lin.zip](https://github.com/squinkylabs/SquinkyVCV/files/7155572/squinkylabs-plug1-1.0.22-lin.zip)

**Comment by brylie** [comment]
Might the issue be related to installing `build-essential` on Ubuntu, which may include updated GNU libc-related libraries?

```
sudo apt install build-essential
```

**Comment by squinkylabs** [comment]
@brylie Anything is possible, but I really think it's the locale thing, with comas being used as decimal points for both of you. And, if you read the above, you can see that I could also get terrible errors if I set my windows machine like that.

So I'm hoping that the build I posted this morning actually fixes all the bugs.

**Comment by burkingman** [comment]
Sorry, that build doesn't work for me. I get the same behaviour I was getting at the start: I can play a note but SFZ Player outputs NaN as soon as I release the button. Nothing in the console output. Here's the log, from the loading of the patch up until the point I closed VCV: 
[sfz_player_bug_log_2021-09-13.txt](https://github.com/squinkylabs/SquinkyVCV/files/7158404/sfz_player_bug_log_2021-09-13.txt)

**Comment by squinkylabs** [comment]
@burkingman that's disappointing. Pretty odd, too. The log looks like it comes from the regular build of my stuff. I'm going to fire up my linux and take a look...

**Comment by squinkylabs** [comment]
ok, here is really dumb idea #1. Hoping this will be it. This is a "real" build, so it won't crash early. 

It does have a lot of debug logging, so if you can capture a log again that would be useful (especially if it doesn't work). When it does work, the last messages you see after patch loading are:
```[3.382 info ./composites/ADSRSampler.h:70] lambda = 0.006250
[3.382 info ./composites/ADSRSampler.h:70] lambda = 0.100000
[3.382 info ./composites/ADSRSampler.h:70] lambda = 0.300000
[3.382 info ./composites/ADSRSampler.h:70] lambda = 0.006250
[3.382 info ./composites/ADSRSampler.h:70] lambda = 0.100000
[3.382 info ./composites/ADSRSampler.h:70] lambda = 0.300000
[3.382 info ./composites/ADSRSampler.h:70] lambda = 0.006250
[3.382 info ./composites/ADSRSampler.h:70] lambda = 0.100000
[3.382 info ./composites/ADSRSampler.h:70] lambda = 0.300000
[3.382 info ./composites/ADSRSampler.h:70] lambda = 0.006250
[3.382 info ./composites/ADSRSampler.h:70] lambda = 0.100000
[3.382 info ./composites/ADSRSampler.h:70] lambda = 0.300000
```

If they are all zero or nan iof inf that would be bad.
[squinkylabs-plug1-1.0.22-lin.zip](https://github.com/squinkylabs/SquinkyVCV/files/7163008/squinkylabs-plug1-1.0.22-lin.zip)

**Comment by burkingman** [comment]
This new version works well on my system. Here's the log, from loading the patch to closing it: 
[sfz_player_bug_log_2021-09-14.zip](https://github.com/squinkylabs/SquinkyVCV/files/7164305/sfz_player_bug_log_2021-09-14.zip)

**Comment by squinkylabs** [comment]
@burkingman : that is very good news. And here is the key section of the log:
```expandAllKV loc=C npunct = true dp=.
is it float
stof after conversion sComa=0,5 fComa=0,500000
stof after conversion sPeriod=0.5 fPeriod=0,000000
ret best coma 0,500000
```
you can see of the first line that I am tryint to ask the computer for the decimal point it's using, but I get the wrong answer (period). I don't now if that's my bug of some linux thing, but I don't care.

next you can see I and getting the release value of `0.5` from the VST. I convert it using both period and coma. Coma gives me `0,5000`, period gives me `0,0000`. I pick the one that is farthest away from zero, and get the correct answer.

This is kind of gross, but will work fine. I'm going to remove all the extra logging and extra locale stuff I'm not using. I'll post another normal build, then hopefully put it in the plugin library.

**Comment by squinkylabs** [comment]
This version should be the same, but without all the logging and some of the unsuccessful code experiments.

[squinkylabs-plug1-1.0.22-lin.zip](https://github.com/squinkylabs/SquinkyVCV/files/7166515/squinkylabs-plug1-1.0.22-lin.zip)

**Comment by burkingman** [comment]
> This is kind of gross, but will work fine.

Sorry you had to do gross things to your code... It does work fine, at least; I hope it does for @brylie as well. Thanks a lot for taking this seriously and putting in the time and effort.

**Comment by squinkylabs** [comment]
@burkingman thank you for all the help. and any "grossness" is of my own making. I can always do something better "later". This one is still quite mysterious to me, but flag it's fixed. btw, if you want to test any new plugins, you are of course more than welcome. Although I do "force" my testers to use Slack, and many people balk at that. In any case if you are interested send me a PM. Now to get ready for a new library build...

**Comment by squinkylabs** [comment]
fixed in 1.0.23
